File: day1/day1.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"log"
	"os"
	"sort"
	"strconv"
	"strings"
)

func main() {
	part1Solution := part1()
	part2Solution := part2()
	fmt.Println("Part 1:", part1Solution)
	fmt.Println("Part 2:", part2Solution)
}

func loadNumbers(filename string) ([]int, []int) {
	lst1 := make([]int, 0)
	lst2 := make([]int, 0)
	file, err := os.Open(filename)
	if err != nil {
		log.Fatal("Failed to open file", filename)
	}
	defer file.Close()
	reader := bufio.NewReader(file)
	for {
		line, err := reader.ReadString('\n')
		if err != nil {
			if err == io.EOF {
				break
			}
			log.Fatal("Failed to read from file")
		}
		split := strings.Fields(line)
		num1, _ := strconv.Atoi(split[0])
		num2, _ := strconv.Atoi(split[1])
		lst1 = append(lst1, num1)
		lst2 = append(lst2, num2)
	}

	return lst1, lst2
}

func abs(x int) int {
	if x >= 0 {
		return x
	} else {
		return -x
	}
}

func part1() int {
	lst1, lst2 := loadNumbers("input.txt")
	sort.Ints(lst1)
	sort.Ints(lst2)

	total := 0
	for i := range len(lst1) {
		diff := lst1[i] - lst2[i]
		total += abs(diff)
	}
	return total
}

func part2() int {
	lst1, lst2 := loadNumbers("input.txt")
	simularityScore := 0
	freqTable := createFrequencyTable(lst2)
	for i := range len(lst1) {
		simularityScore += lst1[i] * freqTable[lst1[i]]
	}
	return simularityScore
}

func createFrequencyTable(slice []int) map[int]int {
	table := make(map[int]int)
	for i := range len(slice) {
		itemCount := count(slice, slice[i])
		table[slice[i]] = itemCount
	}
	return table
}

func count(slice []int, item int) int {
	total := 0
	for i := range len(slice) {
		if slice[i] == item {
			total += 1
		}
	}
	return total
}

File: day1/day1.py
def main() -> None:
    part1_solution = part1()
    print(f"Part 1: {part1_solution}")
    part2_solution = part2()
    print(f"Part 2: {part2_solution}")


def part1() -> int:
    lst1, lst2 = load_numbers("input.txt")
    lst1.sort()
    lst2.sort()

    total = 0
    for i in range(len(lst1)):
        diff = lst1[i] - lst2[i]
        total += abs(diff)
    return total


def part2() -> int:
    lst1, lst2 = load_numbers("input.txt")
    lst1.sort()
    lst2.sort()

    simularity_score = 0
    for num in lst1:
        count = lst2.count(num)
        simularity_score += num * count

    return simularity_score


def load_numbers(file_name: str) -> tuple[list[int], list[int]]:
    lst1 = []
    lst2 = []
    with open(file_name, mode="r") as f:
        for line in f.readlines():
            split_line = line.split()
            lst1.append(int(split_line[0]))
            lst2.append(int(split_line[1]))
    return (lst1, lst2)


if __name__ == "__main__":
    main()
